### DISCUSSION TITLE
Feature Request - Moralis Web3 Plugin

### DISCUSSION THREAD

**Discussion by a-y-u-s-h** [saulpw/visidata#1277] [Ideas]
Hi,

I'm not very familiar with Visidata plugin development yet so I'm suggesting + requesting this plugin. Moralis is like Firebase for Web3 world, but open sourced and exposes database through API keys that can be read in Python. (Reference: [Web3 Data Science](https://www.youtube.com/watch?v=ljHLkuhhN2A)). It would be amazing to have a visidata plugin for this (like reddit plugin) because dapp data analysis could give some really nice insights and help products move smarter.

**Comment by saulpw** [comment]
@a-y-u-s-h Show us some Python code that gets the rows of data you're interested in!  Maybe we can turn it into a plugin right here in the discussion forum :)

**Reply by a-y-u-s-h** [reply]
Yup, the [reference video](https://www.youtube.com/watch?v=ljHLkuhhN2A) demonstrates that at around 8:52.

**Reply by saulpw** [reply]
That's great but unfortunately I can't cut and paste from a video.

**Reply by a-y-u-s-h** [reply]
Oh okay, here it is : 

```py
import pymongo
import pandas as pd


""" 
  ======================================
    IP Address and Port for MongoDB comes
    from Moralis interface, so this could
    be in a config file. The setup for this is mostly in
    Moralis dashboard. Configuration will be different
    for each Moralis project because it's like firebase,
    they have different API keys.
  ======================================
"""
host = "134.289.92.11" 
port = 56728

connection = pymongo.MongoClient(host, port)

""" 
  ======================================
    "parse" is a moralis thing, the table
    called Punks can be fetched like so
  ======================================
"""
data = connection["parse"]["Punks"].find()
dataframe = pd.DataFrame.from_dict(list(data))

""" 
  ======================================
    If we want the second row, or 1st indexed
    row, we can use df.iloc[1]. 
  ======================================
"""
row = dataframe.iloc[1]
```

**Reply by saulpw** [reply]
Okay, so a first pass would look like this (untested code):

```
@VisiData.api
def openurl_mongo(vd, p):
    return MongoSheet(p.name, source=p)

class MongoSheet(Sheet):
    def iterload(self):
        import pymongo
        self.con = pymongo.MongoClient(self.source.host, self.source.port)
        yield from self.con['parse']['Punks'].find()
```

Of course, this won't work if there's not a `Punks` table, I'm guessing.  So probably there should be a MongoIndexSheet that loads the list of tables in the database.  I don't know how to do that, though.
